pkg/storegateway/indexheader/index/postings.go

```diff
@@ -0,0 +1,514 @@
+	// StreamBinaryReader.PostingsOffset only ever calls this method with a single value, although there's a comment
+	// in there about taking advantage of retrieving multiple values at once.
+	PostingsOffset(name string, values ...string) ([]index.Range, error)
+
+	// LabelValues returns a list of values for the label named name that match filter.
+	LabelValues(name string, filter func(string) bool) ([]string, error)
+
+	// LabelNames returns a list of all label names in this table.
+	LabelNames() ([]string, error)
+
+	// ForEachLabelName calls f for each label name in this table.
+	// If f returns a non-nil error for any label name, ForEachLabelName immediately returns that error
+	// and does not call f again.
+	ForEachLabelName(f func(name string) error) error
+
+	// LabelNameCount returns a value which is at least as large as the number of
+	// label names present in this table.
+	LabelNameCount() int
+}
+
+type PostingOffsetTableV1 struct {
+	// For the v1 format, labelname -> labelvalue -> offset.
+	postings map[string]map[string]index.Range
+}
+
+func NewPostingOffsetTable(factory *stream_encoding.DecbufFactory, tableOffset int, indexVersion int, indexLastPostingEnd uint64, postingOffsetsInMemSampling int) (PostingOffsetTable, error) {
+	if indexVersion == index.FormatV1 {
+		return newV1PostingOffsetTable(factory, tableOffset, indexLastPostingEnd)
+	} else if indexVersion == index.FormatV2 {
+		return newV2PostingOffsetTable(factory, tableOffset, indexLastPostingEnd, postingOffsetsInMemSampling)
+	}
+
+	return nil, fmt.Errorf("unknown index version %v", indexVersion)
+}
+
+func newV1PostingOffsetTable(factory *stream_encoding.DecbufFactory, tableOffset int, indexLastPostingEnd uint64) (PostingOffsetTable, error) {
+	t := PostingOffsetTableV1{
+		postings: map[string]map[string]index.Range{},
+	}
+
+	// Earlier V1 formats don't have a sorted postings offset table, so
+	// load the whole offset table into memory.
+	var lastKey []string
+	var prevRng index.Range
+
+	if err := readOffsetTable(factory, tableOffset, func(key []string, off uint64, _ int) error {
+		if len(key) != 2 {
+			return errors.Errorf("unexpected key length for posting table %d", len(key))
+		}
+
+		if lastKey != nil {
+			prevRng.End = int64(off - crc32.Size)
+			t.postings[lastKey[0]][lastKey[1]] = prevRng
+		}
+
+		if _, ok := t.postings[key[0]]; !ok {
+			t.postings[key[0]] = map[string]index.Range{}
+		}
+
+		lastKey = key
+		prevRng = index.Range{Start: int64(off + postingLengthFieldSize)}
+		return nil
+	}); err != nil {
+		return nil, errors.Wrap(err, "read postings table")
+	}
+
+	if lastKey != nil {
+		prevRng.End = int64(indexLastPostingEnd) - crc32.Size
+		t.postings[lastKey[0]][lastKey[1]] = prevRng
+	}
+
+	return &t, nil
+}
+
+func newV2PostingOffsetTable(factory *stream_encoding.DecbufFactory, tableOffset int, indexLastPostingEnd uint64, postingOffsetsInMemSampling int) (PostingOffsetTable, error) {
```

Comment:
Nit: should return a concrete type:

```suggestion
func newV2PostingOffsetTable(factory *stream_encoding.DecbufFactory, tableOffset int, indexLastPostingEnd uint64, postingOffsetsInMemSampling int) (*PostingOffsetTableV1, error) {
```